src/main/java/org/eclipse/uprotocol/streamer/UStreamer.java

```diff
@@ -0,0 +1,128 @@
+     * The input route is the route to listen to and the output route is the route to forward the message to.
+     * The forwarding works by registering a listener on the input route and forwarding the message to the output route.
+     * 
+     * NOTE: Default routing rule (where UAuthority.name is "*") *MUST* be added at the end of the list to avoid overriding other rules.
+     * 
+     * @param in input {@code Route} that the streamer listens to
+     * @param out output {@code Route} that the streamer forwards the message to
+     * @return {@code UStatus} with UCode.OK if the forwarding rule was added successfully,
+     *        {@code UStatus} with UCode.ALREADY_EXISTS if the forwarding rule already exists,
+     *       {@code UStatus} with UCode.INVALID_ARGUMENT if the input and output routes are the same
+     */
+    public UStatus addForwardingRule(Route in, Route out) {
+        Objects.requireNonNull(in, "input cannot be null.");
+        Objects.requireNonNull(out, "output cannot be null.");
+        
+        // Cannot route to itself
+        if (in.equals(out)) {
+            return UStatus.newBuilder().setCode(UCode.INVALID_ARGUMENT).build();
+        }
+
+        // check if the rule already exists in the list
+        if (listeners.stream().anyMatch(p -> p.getInputRoute().equals(in) && p.getOutputRoute().equals(out))) {
+            return UStatus.newBuilder().setCode(UCode.ALREADY_EXISTS).build();
+        }
+
+        TransportListener listener = new TransportListener(in, out);
+        UUri out_uri = UUri.newBuilder().setAuthority(out.getAuthority()).build();
+
+        UStatus result = in.getTransport().registerListener(out_uri, listener);
+
+        if (result.getCode() != UCode.OK) {
+            return result;
+        }
+
+        listeners.add(listener);
+
+        return result;
+    }
+
+    
+    /**
+    * Remove a forwarding rule from the streamer. 
+    * 
+    * @param in input {@code Route} that the streamer listens to
+    * @param out output {@code Route} that the streamer forwards the message to
+    * @return {@code UStatus} with UCode.OK if the forwarding rule was removed successfully,
+    *         {@code UStatus} with UCode.NOT_FOUND if the forwarding rule was not found,
+    *         {@code UStatus} with UCode.INVALID_ARGUMENT if the input and output routes are the same
+    */
+    public UStatus deleteForwardingRule(Route in, Route out) {
+        Objects.requireNonNull(in, "input cannot be null.");
+        Objects.requireNonNull(out, "output cannot be null.");
+        
+        if (in.equals(out)) {
+            return UStatus.newBuilder().setCode(UCode.INVALID_ARGUMENT).build();
+        }
+
+
+        // Unregister the listener with the transport
+        listeners.stream()
+            .filter(p -> p.getInputRoute().equals(in) && p.getOutputRoute().equals(out))
+            .forEach(p -> in.getTransport().unregisterListener(UUri.newBuilder().setAuthority(out.getAuthority()).build(), p));
```

Comment:
what if unregister fails?
can unregister block the thread?
just a feeling in my tummy that this can end up with race conditions

Reply:
Blocking behaviour of the unregister method depends on its implementation 
If one thread is unregistering listener while another is modifying the list, it can lead to race conditions. 
We can use synchronization and an ExecutorService for async operations for added safety.

Reply:
Right now this is only pseudo code and not intended to be used to build an actual streamer.